src/main/java/pl/agh/edu/master_diet/service/BrowserService.java

```diff
@@ -0,0 +1,58 @@
+package pl.agh.edu.master_diet.service;
+
+import pl.agh.edu.master_diet.core.model.database.Browsable;
+import pl.agh.edu.master_diet.repository.BrowserRepository;
+
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+public class BrowserService<T extends Browsable> {
+
+    private static final int ALLOW_SINGLE_TYPO_MINIMUM_WORD_LENGTH = 5;
+
+    private final BrowserRepository<T> browserRepository;
+
+    public BrowserService(BrowserRepository<T> browserRepository) {
+        this.browserRepository = browserRepository;
+    }
+
+    public List<T> search(String searchTerm) {
+        searchTerm = searchTerm.trim().toLowerCase();
+        List<T> result = browserRepository.findBySearchTerm(searchTerm);
+
+        if (searchTerm.length() >= ALLOW_SINGLE_TYPO_MINIMUM_WORD_LENGTH) {
+            List<T> productsForSearchTermWithTypo = searchProductsForTermWithTypo(searchTerm);
+            productsForSearchTermWithTypo.removeAll(result);
+            result.addAll(productsForSearchTermWithTypo);
+        }
+        return result;
+    }
+
+    private List<T> searchProductsForTermWithTypo(String searchTerm) {
+        Set<T> products = new LinkedHashSet<>();
```

Comment:
Does it have to be ```LinkedHashSet```? We sort the results anyway

Reply:
Change to HashSet.